pex/pep_427.py

```diff
@@ -0,0 +1,310 @@
+
+    import attr  # vendor:skip
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class InstallPaths(object):
+
+    CHROOT_STASH = ".prefix"
+
+    @classmethod
+    def chroot(
+        cls,
+        destination,  # type: str
+        project_name,  # type: ProjectName
+    ):
+        # type: (...) -> InstallPaths
+        base = os.path.join(destination, cls.CHROOT_STASH)
+        return InstallPaths(
+            purelib=destination,
+            platlib=destination,
+            headers=os.path.join(base, "include", "site", "pythonX.Y", project_name.raw),
+            scripts=os.path.join(base, "bin"),
+            data=base,
+        )
+
+    @classmethod
+    def interpreter(cls, interpreter):
+        # type: (PythonInterpreter) -> InstallPaths
+        sysconfig_paths = interpreter.identity.paths
+        return InstallPaths(
+            purelib=sysconfig_paths["purelib"],
+            platlib=sysconfig_paths["platlib"],
+            headers=sysconfig_paths["include"],
+            scripts=sysconfig_paths["scripts"],
+            data=sysconfig_paths["data"],
+        )
+
+    purelib = attr.ib()  # type: str
+    platlib = attr.ib()  # type: str
+    headers = attr.ib()  # type: str
+    scripts = attr.ib()  # type: str
+    data = attr.ib()  # type: str
+
+    def __getitem__(self, item):
+        # type: (Text) -> str
+        if "purelib" == item:
+            return self.purelib
+        elif "platlib" == item:
+            return self.platlib
+        elif "headers" == item:
+            return self.headers
+        elif "scripts" == item:
+            return self.scripts
+        elif "data" == item:
+            return self.data
+        raise KeyError("Not a known install path: {item}".format(item=item))
+
+
+def install_wheel_chroot(
+    wheel_path,  # type: str
+    destination,  # type: str
+    compile=False,  # type: bool
+    requested=True,  # type: bool
+):
+    # type: (...) -> InstalledWheel
+
+    metadata_files = install_wheel(
+        wheel_path,
+        InstallPaths.chroot(
+            destination,
+            project_name=ProjectNameAndVersion.from_filename(wheel_path).canonicalized_project_name,
+        ),
+        compile=compile,
+        requested=requested,
+    )
+
+    record_relpath = metadata_files.metadata_file_rel_path("RECORD")
+    assert (
+        record_relpath is not None
+    ), "The {module}.install_wheel function should always create a RECORD.".format(module=__name__)
+
+    direct_url_file_relpath = metadata_files.metadata_file_rel_path("direct_url.json")
+    if direct_url_file_relpath:
+        direct_url_file_abspath = os.path.join(
+            metadata_files.metadata.location, direct_url_file_relpath
+        )
+        if os.path.isfile(direct_url_file_abspath):
+            with open(direct_url_file_abspath) as fp:
+                if urlparse.urlparse(json.load(fp).get("url", "")).scheme == "file":
+                    os.unlink(direct_url_file_abspath)
```

Comment:
This is pre-existing moved code. IIRC it was due to reproducibility issues caused by varying file URLs for the same artifact / local project for, but I'll hunt that down and see if it can't be undone or commented otherwise.

Reply:
I ask because in my reading of the spec it seems like this file is required to exist if the wheel was installed from a URL (e.g., a VCS one) but must not exist if the wheel was installed via a requirement string. This seems like it would violate that. I'm not sure what the consequences are (I guess they are fine, since we've not encountered an issue).

Reply:
Ok, yeah, I was right - this unlink fixed direct_url.json emitted by pip install and undoing this on main causes our reproducibility tests to fail (original PR was #1259):
```
FAILED tests/integration/test_reproducible.py::test_reproducible_build_c_flag_from_source - AssertionError: /tmp/tmpywxi9fe1/pex/tmp/tmpmc2x30mj/2/.deps/my_app-0.0.0-p...
FAILED tests/integration/test_reproducible.py::test_reproducible_build_sdist_requirements - AssertionError: /tmp/tmp1hy72pks/pex/tmp/tmp97rn9ocw/2/.deps/python_crontab...
FAILED tests/integration/test_reproducible.py::test_reproducible_build_c_flag_from_dependency - AssertionError: /tmp/tmpl1mx82lz/pex/tmp/tmpd61jkqcb/2/.deps/future-0.17.1-...
```

But ... that was Pip doing the install and incorrectly emitting a file:// direct URL for a local sdist install. The new Pex install code does not do that and the tests indeed now pass with the unlink commented out; so now removed. Thanks.